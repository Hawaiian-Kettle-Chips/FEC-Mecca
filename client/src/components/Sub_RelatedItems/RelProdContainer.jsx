import React, { useState, useEffect } from 'react';
import axios from 'axios';

import SingleProd from './SingleProd.jsx';

export default function RelProdContainer({
  parentProduct, parentProductID, setParentProductID,
  parentProductStyle, parentReviewMetadata,
  setAsNewOverview, setAllowCardClick,
}) {
  const urls = [
    `/api/products/${parentProductID}/related`,
    
  ];

    axios.get(`/api/products/${parentProductID}/related`)
      .then((response) => {
        const uniqueIds = [...new Set(response.data)];
        setRelatedIDs(uniqueIds);
        return uniqueIds;
      })
      .then((ids) => {
        const productReqs = ids.map((id) => axios.get(`/api/products/${id}`));
        return axios.all(productReqs);
      })
      .then((responses) => {
        const prodSetter = [...responses.map((response) => response.data)];
        setRelatedProds(prodSetter);
        return prodSetter;
      })
      .catch((err) => err);

  // =================== STATES ===================
  const [relatedIDs, setRelatedIDs] = useState([]);

  const [relatedProds, setRelatedProds] = useState([]);

  // =================== EFFECTS ===================

  useEffect(() => getRelatedProdsAndReviews(), [parentProductID]);

  // =================== HELPERS ===================

  const getRelatedProdsAndReviews = () => {
    axios.get(`/api/products/${parentProductID}/related`)
      .then((response) => {
        const uniqueIds = [...new Set(response.data)];
        setRelatedIDs(uniqueIds);
        return uniqueIds;
      })
      .then((ids) => {
        const productReqs = ids.map((id) => axios.get(`/api/products/${id}`));
        return axios.all(productReqs);
      })
      .then((responses) => {
        const prodSetter = [...responses.map((response) => response.data)];
        setRelatedProds(prodSetter);
        return prodSetter;
      })
      .catch((err) => err);
  };

  // =================== COMPONENT ===================
  return (
    <div className="scrollable container">
      {!relatedProds.length && (
        <div className="related-prod null">
          There are no related products to display at this time...
        </div>
      )}
      {relatedProds.map((thisProduct) => (
        <SingleProd
          key={thisProduct.id}
          parentProduct={parentProduct}
          setParentProductID={setParentProductID}
          parentProductStyle={parentProductStyle}
          parentReviewMetadata={parentReviewMetadata}
          thisProduct={thisProduct}
          setAllowCardClick={setAllowCardClick}
          setAsNewOverview={setAsNewOverview}
        />
      ))}
    </div>
  );
}
