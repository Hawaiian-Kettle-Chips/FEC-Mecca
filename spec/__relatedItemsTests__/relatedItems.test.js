import React from 'react';

import {
  cleanup, test, render, screen, fireEevent,
} from '@testing-library/react';

import {
  afterEach, describe, it, expect,
} from '@jest/globals';

import userEvent from '@testing-library/user-event';
import jest from 'jest';

import RelProdContainer from '../../client/src/components/Sub_RelatedItems/RelProdContainer.jsx';

afterEach(cleanup);

describe('Related Items', () => {
  it('renders all related product cards to the DOM', () => {
    const productID = 40348;
    const { container } = (<RelProdContainer productID={productID} />);
    console.log('container: ', container);
    expect(container).toMatchSnapshot();
  });
});

// RelProdContainer.jsx

// getRelatedProdsAndReviews should return a response data
//   array length 7 when productID state is 40348

// describe('Related Products Container', () => {
//   test('response data is an array of length 7 when productID is 40348', () => {
//     const id = 40348;
//     return getRelatedProdsAndReviews()
//       .then((response) => {
//         expect(response.data).toHaveLength(7);
//       });
//     // const {container} = render(<RelProdContainer productID={40348})
//   });
// });

// the relatedIDs state should not have duplicate IDs
//   the relatedIDs state should be length 6 when productID state is 40348

// the relatedProds state should be an array of same length as relatedIDs state

// there should be the same # of 'SingleProd' components rendered as relatedProds' length
