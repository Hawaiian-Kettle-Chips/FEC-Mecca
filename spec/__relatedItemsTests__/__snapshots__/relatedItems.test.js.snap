// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Related Items renders all related product cards to the DOM 1`] = `undefined`;
